### PR TITLE
@pavelvasev merge: Merge recursive props

### PR DESCRIPTION
**Pull request by henrikrudstrom**
merged recursive properties from @pavelvasev branch
@ChALkeR

### CODE REVIEW

File: src/qtcore/qml/QMLEngine.js [hunk 1]
```diff
@@ -478,8 +478,8 @@ QMLEngine = function (element, options) {
             var property = this.bindedProperties[i];
             if (!property.binding)
               continue; // Probably, the binding was overwritten by an explicit value. Ignore.
-            property.binding.compile();
-            property.update();
+            if (property.needsUpdate)
++               property.update();
```
Comment (ChALkeR) [inline src/qtcore/qml/QMLEngine.js]: Stray `+`.

**Comment by ChALkeR** [comment]
@henrikrudstrom Thanks! Will take a look at this soon.
I will probably just force-push this branch to fix minor nits, ok?

**Comment by henrikrudstrom** [comment]
Thats totally fine!
Fri 6 May 2016 kl. 17:00 skrev Сковорода Никита Андреевич <
notifications@github.com>:

**Comment by ChALkeR** [comment]
https://github.com/qmlweb/qmlweb/commit/595a29fbf1ebca1b2aa9cc8d6bd7d61a4cd78166 got missed, I will add that on top.

File: src/qtcore/qml/QMLEngine.js (hunk 1 above)
Comment (ChALkeR) [inline src/qtcore/qml/QMLEngine.js]: I better not forget removing this =).

**Comment by ChALkeR** [comment]
I am going to rename `evaluatingPropertyTop` back to to `evaluatingProperty`, that was unneeded change.

**Comment by ChALkeR** [comment]
Btw, this merges 0aab767, 09b8243, 0e80a6e, 595a29f.

File: src/qtcore/qml/qml.js [hunk 2]
```diff
@@ -306,6 +308,11 @@ function applyProperties(metaObject, item, objectScope, componentScope) {
     _executionContext = componentScope;
     for (i in metaObject) {
         var value = metaObject[i];
+        if (i == "id" || i == "$class") { // keep them
+          item[i] = value;
+          continue;
+        }
```
Comment (ChALkeR) [inline src/qtcore/qml/qml.js]: @pavelvasev, what's going on here? This contradicts the check below, `i == 'id'` below never fires.

File: src/qtcore/qml/qml.js [hunk 3]
```diff
@@ -4,6 +4,9 @@ var GETTER = "__defineGetter__",
     // which property called the getter of a certain other property for
     // evaluation and is thus dependant on it.
     evaluatingProperty = undefined,
+    evaluatingPropertyStack = [],
+    evaluatingPropertyPaused = false,
```
Comment (ChALkeR) [inline src/qtcore/qml/qml.js]: Unused. Do we need this? /cc @pavelvasev

File: src/qtcore/qml/QMLEngine.js [hunk 4]
```diff
@@ -474,23 +475,33 @@ QMLEngine = function (element, options) {
...
+        // we use `while`, because $initializePropertyBindings may be called recursive (because of Loader and/or createQmlObject )
+        while (this.bindedProperties.length > 0) {
+            var property = this.bindedProperties.shift();
             if (!property.binding)
               continue; // Probably, the binding was overwritten by an explicit value. Ignore.
-            property.binding.compile();
-            property.update();
+            if (property.needsUpdate)
+                property.update();
+            else if (["width","height","fill","x","y","left","right","top","bottom"].indexOf(property.name) >= 0) {
```
Comment (ChALkeR) [inline src/qtcore/qml/QMLEngine.js]: Why are those special? What about anchors? /cc @pavelvasev

**Comment by ChALkeR** [comment]
Ah, and tracking: #32.

**Comment by ChALkeR** [comment]
Rebased.

File: src/qtcore/qml/qml.js [hunk 5]
```diff
@@ -2,6 +2,9 @@
 // which property called the getter of a certain other property for
 // evaluation and is thus dependant on it.
 var evaluatingProperty;
+var evaluatingPropertyStack = [];
+var evaluatingPropertyPaused = false;
```
Comment (ChALkeR) [inline src/qtcore/qml/qml.js]: This is unused.

**Comment by ChALkeR** [comment]
Ok, rebase didn't work, reverting rebase =).

**Comment by ChALkeR** [comment]
Merging #248 broke this.

**Comment by ChALkeR** [comment]
Ok, I fixed the rebase.

The last two commits LGTM, the first three seem good, I will just merge them for now — we can fix things later after the merges are complete =).

**Comment by stephenmdangelo** [comment]
@ChALkeR So, I've run into a use-case where this breaks geometry updating, as the recursion checks within `update*Geometry` cause the methods to return before properly applying the appropriate updates.  Reverting these commits fixes it.  I haven't had a chance to pin-point it directly, yet.

I'm guessing it's related to the forced `update*Geometry` calls at the end of `$initializePropertyBindings`.  It seems to only call them if they are internal connections between those geometry-related property names and the geometry of the item with those properties, which won't take into account places where anchors or geometry-related properties refer to properties on other items.

**Pull request merged**